bundle/config/mutator/run_as.go

```diff
@@ -23,12 +27,120 @@ func (m *setRunAs) Name() string {
 	return "SetRunAs"
 }
 
+// Resources that satisfy one of the following conditions:
+//  1. Allow to set run_as for the resources to a different user from the current
+//     deployment user. For example, jobs.
+//  2. Does not make sense for these resources to run_as a different user. We do not
+//     have plans to add platform side support for `run_as` for these resources.
+//     For example, experiments or registered models.
+var allowListForRunAsOther = []string{"jobs", "models", "registered_models", "experiments"}
+
+// Resources that do not allow setting a run_as identity to a different user but
+// have plans to add platform side support for `run_as` for these resources at
+// some point in the future. For example, pipelines, model serving endpoints or lakeview dashboards.
+//
+// We expect the allow list and the deny list to form mutually exclusive and exhaustive
+// sets of all resource types that are supported by DABs.
+var denyListForRunAsOther = []string{"pipelines", "model_serving_endpoints"}
+
+type errorUnsupportedResourceTypeForRunAs struct {
+	resourceType  string
+	resourceValue dyn.Value
+	currentUser   string
+	runAsUser     string
+}
+
+// TODO(6 March 2024): This error message is big. We should split this once
+// diag.Diagnostics is ready.
+// TODO(6 March 2024): Link the docs page describing run_as semantics in the error below
+// once the page is ready.
+func (e errorUnsupportedResourceTypeForRunAs) Error() string {
+	return fmt.Sprintf("%s are not supported when the current deployment user is different from the bundle's run_as identity. Please deploy as the run_as identity. List of supported resources: [%s]. Location of the unsupported resource: %s. Current identity: %s. Run as identity: %s", e.resourceType, strings.Join(allowListForRunAsOther, ", "), e.resourceValue.Location(), e.currentUser, e.runAsUser)
+}
+
+func getRunAsIdentity(runAs dyn.Value) (string, error) {
```

Comment:
Not sure if we really need dynamic value reading here instead of just few lines of code like

```
if b.Config.Bundle.RunAs.ServicePrincipalName != "" {
  return  b.Config.Bundle.RunAs.ServicePrincipalName
}

if b.Config.Bundle.RunAs.UserName != "" {
  return  b.Config.Bundle.RunAs.ServicePrincipalName
}

Reply:
Well, we should ensure both of them are not set at the same time, but that can also be done using the static types.

The main benefit is the location information in the error message, which is not possible with static types.

Reply:
For me it seems a bit overly complex to have all this code just to get location information in the error, especially if we will do this everytime we need the location.

Shall we introduce `b.Config.GetLocationByPath(path string)` to get the location instead? Then we could use static constructs here everywhere and just call `b.Config.GetLocationByPath` in the error.

Thoughts?

Reply:
Added a `TryLocation` function to `bundle.Config` and using that here. We can however no longer differentiate between the `run_as.user_name` (or service_principal_name) not being set and it being set to the empty string.

Reply:
> Are you working with Julia to define a blurb for this one as well?
Yes, Julia is picking up the docs page for the run_as functionality.